### PR TITLE
FLAG-61: Fixed issues related to adding/editing/deleting flag priorities

### PR DESCRIPTION
**Pull request by druchniewicz**
Related ticket: https://issues.openmrs.org/browse/FLAG-61

What changes I made:
- fixed validation when we want to save priority with already existing name (name must be unique). Before fix we got an error page with exception message (org.hibernate.exception.ConstraintViolationException: could not execute statement)
- fixed validation when we want to save priority with empty 'style' value. This field is mandatory and before fix we got an error page with exception message (org.hibernate.PropertyValueException: not-null property references a null or transient value : org.openmrs.module.patientflags.Priority.style)
- added validation message related to missing style value. Style value is mandatory and before fix we were not able to see validation message
- removed duplicated validation messages. Before fix when we want to save priority for example with empty name we got validation messages 'Priority name is required' twice
- fixed removing priority even though it is not associated with any flag. When we want to remove priority associated with any flag we are not able to do it what makes sense because there is connection between them, but when priority is not connected with any flag we should be able to remove priority. Before fix we got an error page with exception message (Request processing failed; nested exception is java.lang.NullPointerException)

### CODE REVIEW

File: omod/src/main/java/org/openmrs/module/patientflags/web/validators/PriorityValidator.java [hunk 1]
```diff
@@ -13,33 +13,47 @@
...
+import org.openmrs.module.patientflags.api.FlagService;
 import org.springframework.validation.Errors;
 import org.springframework.validation.ValidationUtils;
 
 /**
  * Validator for the Priority class
  */
 public class PriorityValidator {
-	
+
```
Comment (HerbertYiga) [inline omod/src/main/java/org/openmrs/module/patientflags/web/validators/PriorityValidator.java:26]: thanks @druchniewicz  we can remove the unwanted spaces.,have tested the changes on a running instance ?

Comment (druchniewicz) [reply omod/src/main/java/org/openmrs/module/patientflags/web/validators/PriorityValidator.java:26]: Hi @HerbertYiga , yes, changes have been tested on a running instance.
regarding unwanted spaces what exactly do you mean? should I re-indent entire file to remove all tabulations and change it to e.g. 4 spaces?

File: api/src/main/resources/messages.properties [hunk 2]
```diff
@@ -96,7 +96,9 @@ patientflags.errors.flagNameExists=A flag with this name already exists
 patientflags.errors.tagNameExists=A tag with this name already exists
 patientflags.errors.noPriority=Priority is required
 patientflags.errors.noPriorityName=Priority name is required
+patientflags.errors.noStyleName=Priority style is required
```
Comment (kdaud) [inline api/src/main/resources/messages.properties:99]: @druchniewicz did you intend to append the string `^M` on the property?

**Comment by kdaud** [comment]
Well done @druchniewicz !

Could you let me know the command you're using to run the module so that I test these changes on actual server?

File: api/src/main/resources/messages.properties (hunk 2 above)
Comment (druchniewicz) [reply api/src/main/resources/messages.properties]: no, it was not intended, it is related to different end-line characters on different systems and I haven't seen it during development. I already fixed it

**Comment by druchniewicz** [comment]
> Hi @kdaud 
I use 'mvn clean install' to build omod file

**Comment by kdaud** [comment]
Will test this tonight and will feedback here on the findings!

**Comment by kdaud** [comment]
@druchniewicz  checkout my comment on the ticket!

File: api/src/main/java/org/openmrs/module/patientflags/api/FlagService.java [hunk 3]
```diff
@@ -407,4 +407,13 @@ public interface FlagService extends OpenmrsService {
...
 	public Collection<Privilege> getPrivileges();
+
+	/**
+	 * Checks if given priority's name is duplicated
+	 *
+	 * @param priority object to verify uniqueness of priority name
+	 * @return true or false depending on priority's name is duplicated or not
+	 */
+	@Authorized(value = { "Manage Flags" })
+	boolean isPriorityNameDuplicated(Priority priority);
```
Comment (kdaud) [inline api/src/main/java/org/openmrs/module/patientflags/api/FlagService.java:418]: I think we can make this function a public one.

File: api/src/main/java/org/openmrs/module/patientflags/db/FlagDAO.java [hunk 4]
```diff
@@ -228,4 +228,12 @@ public interface FlagDAO {
...
 	 */
 	public void purgeDisplayPoint(Integer displayPointId) throws DAOException;
+
+	/**
+	 * Checks if given priority's name is duplicated
+	 *
+	 * @param priority object to verify uniqueness of priority name
+	 * @return true or false depending on priority's name is duplicated or not
+	 */
+	boolean isPriorityNameDuplicated(Priority priority);
```
Comment (kdaud) [inline api/src/main/java/org/openmrs/module/patientflags/db/FlagDAO.java:238]: Refer to my comment [above](https://github.com/openmrs/openmrs-module-patientflags/pull/36/files#r865871231)

File: api/src/main/java/org/openmrs/module/patientflags/db/hibernate/HibernateFlagDAO.java [hunk 5]
```diff
@@ -314,4 +315,25 @@ public void purgeDisplayPoint(Integer displayPointId) throws DAOException {
 		DisplayPoint displayPoint = getDisplayPoint(displayPointId);
 		sessionFactory.getCurrentSession().delete(displayPoint);
 	}
+
+	public boolean isPriorityNameDuplicated(Priority priority) {
+		Criteria criteria = sessionFactory.getCurrentSession().createCriteria(Priority.class);
+		addEqualsRestriction(criteria, "name", priority.getName());
+		addNotEqualsRestriction(criteria, "priorityId", priority.getPriorityId());
+		addEqualsRestriction(criteria, "retired", false);
```
Comment (kdaud) [inline api/src/main/java/org/openmrs/module/patientflags/db/hibernate/HibernateFlagDAO.java:323]: Could you walk me through what code line 323 is performing?

File: api/src/main/resources/messages.properties [hunk 6]
```diff
@@ -1,102 +1,92 @@
...
+patientflags.errors.messageTooLong=Message can only be 255 characters in length
+patientflags.errors.tagTooLong=Tag can only be 255 characters in length
+patientflags.errors.flagNameExists=A flag with this name already exists
+patientflags.errors.tagNameExists=A tag with this name already exists
+patientflags.errors.noPriority=Priority is required
+patientflags.errors.noPriorityName=Priority name is required
+patientflags.errors.noStyleName=Priority style is required
+patientflags.errors.noRank=Rank is required
+patientflags.errors.priorityNameTooLong=Name can only be 255 characters in length
+patientflags.errors.noUniqueName=Priority name already exists. Change name to unique
```
Comment (kdaud) [inline api/src/main/resources/messages.properties:90]: We can improve the alert message in particular for the last part of it(`Change name to unique`). It seems to suggest the name should be changed to **unique**

**Comment by kdaud** [comment]
Great work @druchniewicz ! I have tested out the changes and the solution is working out though have noticed a few things that have highlighted below:

1. The rank field is accepting values that are negative. Not sure though but would expect to ranks to be a positive integer. Besides that the field is also accepting non-interger values for-example (6kpm) which should not be. I think we can add validation to input field to catch those exceptions.
2. Am wondering whether for the filed of Style we can have a drop down list where a user just need to select incase these styles are well known.
3. Lastly, If we can add notifications that can show up when the data saved/deleted/edited successfully would be great.

File: api/src/main/java/org/openmrs/module/patientflags/api/FlagService.java (hunk 3 above)
Comment (druchniewicz) [reply api/src/main/java/org/openmrs/module/patientflags/api/FlagService.java:418]: This is interface method so it is public by default

File: api/src/main/resources/messages.properties (hunk 6 above)
Comment (druchniewicz) [reply api/src/main/resources/messages.properties:90]: So what exactly message should be? Name should be changed to unique that's why I added such message

**Comment by druchniewicz** [comment]
> Great work @druchniewicz ! Apart from putting few comments above to better the work, I have tested out the changes and the solution is working out +1 though have noticed a few things that have highlighted below:
> 
> 1. The rank field is accepting values that are negative. Not sure though but would expect to ranks to be a positive integer. Besides that the field is also accepting non-interger values for-example (6kpm) which should not be. I think we can add validation to input field to catch those exceptions.
> 2. Am wondering whether for the filed of Style we can have a drop down list where a user just need to select incase these styles are well known.
> 3. Lastly, If we can add notifications that can show up when the data saved/deleted/edited successfully would be great.

I think those remarks are fine but I would suggest to create a new ticket for these improvements (I can do it). I just wanted to fix all required things that caused issues (bug fixing).

File: api/src/main/resources/messages.properties (hunk 6 above)
Comment (kdaud) [reply api/src/main/resources/messages.properties:90]: > Priority name already exists. Change name to unique

The point here is that the user is likely going to type the name `unique` as a priority name based on the alert message.

How about if we use `Priority name already exists. Change to another name`?

Comment (druchniewicz) [reply api/src/main/resources/messages.properties:90]: ah ok I see your point that user can try to type 'unique' as a name. ok I will change message into 'Priority name already exists. Change to another name'

Comment (druchniewicz) [reply api/src/main/resources/messages.properties]: I've already fixed it

**Comment by druchniewicz** [comment]
> > I think those remarks are fine but I would suggest to create a new ticket for these improvements (I can do it). I just wanted to fix all required things that caused issues (bug fixing).
> 
> @druchniewicz it's fair enough!

@kdaud I've alredy created ticket for these improvements - https://issues.openmrs.org/browse/FLAG-62

**Review by kdaud — APPROVED** [review]
LGTM

**Comment by kdaud** [comment]
@HerbertYiga do you have any more review before we have this guy merged?

**Pull request merged**